spigot/src/main/kotlin/com/mineinabyss/geary/minecraft/dsl/GearyExtension.kt

```diff
@@ -24,147 +26,181 @@ import org.reflections.util.FilterBuilder
+    /*public fun autoscanSingletonSystems(init: AutoScanner.() -> Unit = {}) {
+        val scanner = AutoScanner().apply(init)
+        scanner.registerSerializers<TickingSystem, GearySystem> { kClass, _ ->
+            kClass.objectInstance?.let { system(it) }
+        }
+    }*/
+
     /**
-     * Scans for [GearyComponent]s in the classpath of the provided [plugin]s [ClassLoader]. Optional
-     * [componentPath] can be used to restrict what packages are scanned.
+     * Registers serializers for any type [T] on the classpath of [plugin]'s [ClassLoader].
      *
-     * _Note that if the plugin is loaded using a custom classloading solution, autoscan may not work._
+     * @see AutoScanner
      */
     @InternalSerializationApi
-    public fun autoscanComponents(plugin: Plugin, componentPath: String? = null) {
-        autoscan(
-            plugin,
-            componentPath
-        ) { kClass: KClass<GearyComponent>,
-            kSerializer: KSerializer<GearyComponent>,
-            module: PolymorphicModuleBuilder<GearyComponent> ->
-            module.component(
-                kClass,
-                kSerializer,
-            )
-        }
+    public inline fun <reified T : Any> autoscan(
+        init: AutoScanner.() -> Unit = {},
+        noinline addSubclass: SerializerRegistry<T> = { kClass, serializer -> subclass(kClass, serializer) }
+    ) {
+        AutoScanner().apply(init).registerSerializers(T::class, addSubclass)
+    }
+
+    private companion object {
+        private data class CacheKey(val plugin: Plugin, val path: String?, val excluded: Collection<String>)
+
+        private val reflectionsCache = mutableMapOf<CacheKey, Reflections>()
     }
 
     /**
-     * Scans for [GearyAction]s in the classpath of the provided [plugin]s [ClassLoader]. Optional
-     * [actionPath] can be used to restrict what packages are scanned.
+     * DSL for configuring automatic scanning of classes to be registered into Geary's [SerializersModule].
+     *
+     * A [path] to limit search to may be specified. Specific packages can also be excluded with [excludePath].
+     * Annotate a class with [ExcludeAutoscan] to exclude it from automatically being registered.
      *
      * _Note that if the plugin is loaded using a custom classloading solution, autoscan may not work._
+     *
+     * @property path Optional path to restrict what packages are scanned.
+     * @property excluded Excluded paths under [path].
      */
-    @InternalSerializationApi
-    public fun autoscanActions(plugin: Plugin, actionPath: String? = null) {
-        autoscan(
-            plugin,
-            actionPath
-        ) { kClass: KClass<GearyAction>,
-            kSerializer: KSerializer<GearyAction>,
-            module: PolymorphicModuleBuilder<GearyAction> ->
-            module.action(
-                kClass,
-                kSerializer
+    @GearyExtensionDSL
+    public inner class AutoScanner {
```

Comment:
Does this need to be public.

Reply:
Needed for inline autoscan function that uses a reified type